java-checks/src/main/java/org/sonar/java/checks/TryWithResourcesCheck.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import org.sonar.api.server.rule.RulesDefinition;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.java.JavaVersionAwareVisitor;
+import org.sonar.java.tag.Tag;
+import org.sonar.plugins.java.api.JavaVersion;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TryStatementTree;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+
+@Rule(
+  key = "S2093",
+  name = "Try-with-resources should be used",
+  priority = Priority.MAJOR,
+  tags = {Tag.PITFALL})
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.LOGIC_RELIABILITY)
+@SqaleConstantRemediation("15min")
+@ActivatedByDefault
+public class TryWithResourcesCheck extends SubscriptionBaseVisitor implements JavaVersionAwareVisitor {
+
+  private final Deque<TryStatementTree> withinTry = new LinkedList<>();
+  private final List<TryStatementTree> reported = new ArrayList<>();
+
+  @Override
+  protected void scanTree(Tree tree) {
+    super.scanTree(tree);
+    withinTry.clear();
+    reported.clear();
+  }
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.TRY_STATEMENT, Tree.Kind.NEW_CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (tree.is(Tree.Kind.TRY_STATEMENT)) {
+      withinTry.push((TryStatementTree) tree);
+    } else if (withinStandardTry() && ((NewClassTree) tree).symbolType().isSubtypeOf("java.lang.AutoCloseable") && !reported.contains(withinTry.peek())) {
+      reported.add(withinTry.peek());
+      reportIssue(withinTry.peek().tryKeyword(), "Change this \"try\" to a try-with-resources." + context.getJavaVersion().java7CompatibilityMessage());
```

Comment:
@benzonico with the current implementation it is not clear what is the cause of the issue. We should probably create one issue per instantiated resource with a message like `Change the enclosing "try" to a try-with-resources.`